### ISSUE TITLE
akka-http RC4 fails under load with 'Connection reset by peer' (regression)

### ISSUE THREAD

**Issue by mo-gr** [akka/akka#17854]
The following program works like a charm with akka-http RC3

    object Main extends App {
      implicit val actorSystem: ActorSystem = ActorSystem("main-actor-system")
      implicit val actorFlowMaterializer = ActorFlowMaterializer()
      val bindingFuture = Http(actorSystem).bindAndHandle((r: RequestContext) => r.complete("ok"), "0.0.0.0", 8080)
    }

When I hit it with apache bench like like so 

    ab -c 1 -n 1000 http://localhost:8080/

However, when I upgrade to RC4 and change the ActorFlowMaterializer to ActorMaterializer, I get `apr_socket_recv: Connection reset by peer (54)` from apache bench after something between 2 and 100 requests.

I have no idea, what could be the issue...

Me environment:
- java version "1.8.0_45"
- MacOS 10.10.3
- scala 2.11.6

The output from akka while failing is:

```
[info] Loading project definition from /Users/mgrauel/Coding/scompositor/project
[info] Set current project to scompositor (in build file:/Users/mgrauel/Coding/scompositor/)
[info] Running foo.Main 
[INFO] [06/26/2015 17:09:26.340] [main-actor-system-akka.actor.default-dispatcher-16] [akka://main-actor-system/system/IO-TCP/selectors/$a/3] Message [akka.io.SelectionHandler$ChannelReadable$] from Actor[akka://main-actor-system/deadLetters] to Actor[akka://main-actor-system/system/IO-TCP/selectors/$a/3#-658865784] was not delivered. [1] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [06/26/2015 17:09:26.483] [main-actor-system-akka.actor.default-dispatcher-14] [akka://main-actor-system/system/IO-TCP/selectors/$a/9] Message [akka.io.SelectionHandler$ChannelReadable$] from Actor[akka://main-actor-system/deadLetters] to Actor[akka://main-actor-system/system/IO-TCP/selectors/$a/9#-237500142] was not delivered. [2] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [06/26/2015 17:09:26.483] [main-actor-system-akka.actor.default-dispatcher-14] [akka://main-actor-system/system/IO-TCP/selectors/$a/9] Message [akka.io.SelectionHandler$ChannelReadable$] from Actor[akka://main-actor-system/deadLetters] to Actor[akka://main-actor-system/system/IO-TCP/selectors/$a/9#-237500142] was not delivered. [3] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [06/26/2015 17:12:13.397] [main-actor-system-akka.actor.default-dispatcher-26] [akka://main-actor-system/system/IO-TCP/selectors/$a/11] Message [akka.io.SelectionHandler$ChannelReadable$] from Actor[akka://main-actor-system/deadLetters] to Actor[akka://main-actor-system/system/IO-TCP/selectors/$a/11#-1314876149] was not delivered. [4] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [06/26/2015 17:12:13.423] [main-actor-system-akka.actor.default-dispatcher-26] [akka://main-actor-system/system/IO-TCP/selectors/$a/12] Message [akka.io.SelectionHandler$ChannelReadable$] from Actor[akka://main-actor-system/deadLetters] to Actor[akka://main-actor-system/system/IO-TCP/selectors/$a/12#588690784] was not delivered. [5] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

**Comment by drewhk** [comment]
I have a suspicion. Let's try to reproduce it first, then we should retry again by turning on halfClose again on the server side. I think the akka.io actor underneath does not properly flush on Close and hence terminates the connection before the bytes has been written (although the stream actor waits for the Ack on the last write).

**Comment by drewhk** [comment]
In fact I think this is the same issue that causes https://github.com/akka/akka/issues/17804
Unfortunately this can be only fixed by a new Akka 2.3.13 release, which complicates things.

**Comment by mo-gr** [comment]
Ok. Thanks for clarification. Looking forward to a fix

**Comment by drewhk** [comment]
It is easy to test this, since if the halfClose is turned on for the server connection I would expect this to work.

**Comment by viktorklang** [comment]
Do we have a regression test?

**Comment by drewhk** [comment]
We have some, but it seems they were not enough. I think it is related to
the failing TcpSpec case, but not 100% sure. We will add some more
stress-tests.

-Endre

On Mon, Jun 29, 2015 at 10:52 AM, Viktor Klang (√) <notifications@github.com

**Comment by viktorklang** [comment]
@drewhk I could reproduce that one fairly easily on my box

**Comment by drewhk** [comment]
On Mon, Jun 29, 2015 at 11:17 AM, Viktor Klang (√) <notifications@github.com
> wrote:

> @drewhk <https://github.com/drewhk> I could reproduce that one fairly
> easily on my box
>

But that's the thing, it didn't fail on my machine ever, and the PR
validation didn't fail when it was merged. And we are not yet sure that
this ticket is actually related to the TcpSpec, it is just my current
assumption.

-Endre

**Comment by drewhk** [comment]
Hm, the undelying akka.io actor does flush properly it seems (at least there is code to do exactly that), so this might be related to something else.

**Comment by jrudolph** [comment]
It somewhat reproduces for me. I get some dead letters. What I see is that on some connections the infamous HTTP connection closure RST (explained in [a section of RFC 7230](http://tools.ietf.org/html/rfc7230#section-6.6)) occurs:

```
13:26:49.474293 IP6 localhost.37655 > localhost.http-alt: Flags [S], seq 2195144125, win 43690, options [mss 65476,sackOK,TS val 12540091 ecr 0,nop,wscale 10], length 0
13:26:49.474300 IP6 localhost.http-alt > localhost.37655: Flags [S.], seq 1350759897, ack 2195144126, win 43690, options [mss 65476,sackOK,TS val 12540091 ecr 12540091,nop,wscale 10], length 0
13:26:49.474307 IP6 localhost.37655 > localhost.http-alt: Flags [.], ack 1, win 43, options [nop,nop,TS val 12540091 ecr 12540091], length 0
13:26:49.474320 IP6 localhost.37655 > localhost.http-alt: Flags [P.], seq 1:83, ack 1, win 43, options [nop,nop,TS val 12540091 ecr 12540091], length 82
13:26:49.474328 IP6 localhost.http-alt > localhost.37655: Flags [.], ack 83, win 43, options [nop,nop,TS val 12540091 ecr 12540091], length 0
13:26:49.479764 IP6 localhost.http-alt > localhost.37655: Flags [P.], seq 1:164, ack 83, win 43, options [nop,nop,TS val 12540093 ecr 12540091], length 163
13:26:49.479796 IP6 localhost.37655 > localhost.http-alt: Flags [.], ack 164, win 44, options [nop,nop,TS val 12540093 ecr 12540093], length 0
13:26:49.479873 IP6 localhost.http-alt > localhost.37655: Flags [F.], seq 164, ack 83, win 43, options [nop,nop,TS val 12540093 ecr 12540093], length 0
13:26:49.480499 IP6 localhost.http-alt > localhost.37655: Flags [R.], seq 165, ack 83, win 43, options [nop,nop,TS val 12540093 ecr 12540093], length 0
```

It seems the HTTP implementation currently full-closes the connection after having passed the response to the kernel. However, this happens so fast, that the client hasn't even received the response or at least wasn't able to ACK the last packet of the response. When the kernel on the server side receives the ACK, the socket is already gone and reception of the ACK is answered with a RST. In the case from the dump the client may have received all data, but the same would happen even if the server had data for the client still in its send buffer, in-flight, or in the client's receive buffer, in which case data would be lost.

As @drewhk suggested, changing to `halfClose = true` "fixes" this issue. However, this fix cannot be enough, as we need to guard the server against keeping connections open (i.e. in FIN_WAIT_1 or FIN_WAIT_2) where the client silently went away which the server cannot possibly detect. 

See https://httpd.apache.org/docs/2.0/misc/fin_wait_2.html for more information.

As this can be a problem for all kinds of TCP-based protocols it would maybe make sense to think about another close mode where the TCP (stream) layer half-closes the connection and waits for the peer to close its side of the connection for some while before closing it finally and releasing resources. The above apache link mentions that SO_LINGER meant to help for these cases but isn't working consistently enough to replace a custom timeout.

**Comment by jrudolph** [comment]
To sum up:
 * with `halfClose = false` (RC4 behavior), it is likely that the client cannot read the complete response
 * with `halfClose = true` (RC3 behavior), we may leak connections in the `FIN_WAIT` states if the client silently goes away
 * with a possbile `closeMode = LingerFor(10.seconds)`, we let the client drain our send buffer for another 10 seconds before we cut it off.

It would be nice to have in Akka IO (we could then more easily fix spray/spray#1058) but could probably also be implemented directly in TCP streams for this release.

**Comment by viktorklang** [comment]
@jrudolph So, for my understanding, am I correct in my interpretation that TCP does not buffer the FIN and sends it out-of-band even if there are pending acks?

**Comment by jrudolph** [comment]
@viktorklang Yes, you are right, actually this shouldn't happen when you use `close` as explained in "Case 1" of [UNIX Socket FAQ
» 4.6 - What exactly does SO_LINGER do?](http://developerweb.net/viewtopic.php?id=2982), so now I'm a bit confused myself.

It seems all of my comments above would only matter for persistent connections (which this isn't about because `ab` sends non-persistent `HTTP/1.0` requests).

**Comment by drewhk** [comment]
Well, the Tcp actor does flush properly after Close, and I only send the Close from the stream side once the last write has been actually Acked. What I found though, that the TCP actor does not unwatch the user side actor after Close has been received, so there can be a DeathPactException actually as far as I see. Btw, there was also a bug in the stream side which sent an Abort when it shouldn't but I still have test failures after fixing that.

**Comment by viktorklang** [comment]
@drewhk does fixing the unwatch impact the reproducibility?

**Comment by drewhk** [comment]
I have no clue yet, because I can only reproduce on Jenkins, and streams is built from published artifacts. So I try to put together a test-case on 2.3 branch directly (no streams) and see if it happens on Jenkins.

**Comment by jrudolph** [comment]
You can also use https://github.com/spray/akka/tree/w/merged-with-2.3 which is release-2.3 and release-2.3-dev somewhat cleanly merged with project dependencies. Not sure if PR validation works for it, though.

**Comment by jrudolph** [comment]
What happens if the inbound end of the TCP stream is cancelled before the outbound is completed in `halfClose = false` mode? From what I see, the TCP stream sends mostly `ConfirmedClose` with some `Close` in between. Could this be a race between outbound "Complete" and inbound "Cancel" where the outbound "Complete" wins most of the time for some reason?

**Comment by drewhk** [comment]
Cancelling the inbound side does not affect anything. You mean when the remote write side half-closes? Btw, there is definitely a race between Terminated in the akka.io actors, but there are many things happening here at the same time, and I want to first fix the TcpSpec that tends to fail, and which is not exactly same as the error reported in this ticket.

Btw, as I said there was at least one bug with an abort, but I am hunting for something else, too.

**Comment by jrudolph** [comment]
I added this in `tcpOutputs.complete` to trace the decision why `ConfirmedClose` or `Close` is used:

```scala
println(s"tcpInputs.isClosed: ${tcpInputs.isClosed} halfClose: $halfClose lastWriteAcked: $lastWriteAcked")
if (tcpInputs.isClosed && (halfClose || lastWriteAcked)) {
```

The output looks like this:

```
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: true
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: true
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
tcpInputs.isClosed: true halfClose: false lastWriteAcked: false
```

So `tcpInputs.isClosed` was always `true` even when I made sure that the input side was never cancelled! Which close mode is used then only depends on `lastWriteAcked`. This is because there's also this snippet in `writePump.pumpFinished` which always calls `tcpInputs.cancel()` before calling `tcpOutputs.complete`:

```scala
    override protected def pumpFinished(): Unit = {
      if (fullClose) {
        // In full close mode we shut down the read size immediately once the write side is finished
        tcpInputs.cancel()
        primaryOutputs.complete()
        readPump.pump()
      }
      tcpOutputs.complete()
      primaryInputs.cancel()
      tryShutdown()
    }
```
I think there may be several bugs:
 * in HTTP we need to make sure not to cancel the incoming side
 * this bug here with `fullClose` handling
 * something else I saw, where stream TCP always sends both `ConfirmedClose` and later `Close` (the latter one being ignored)

**Comment by drewhk** [comment]
> So tcpInputs.isClosed was always true even when I made sure that the input side was never cancelled! 

WDYM? It can be only set to true if the cancel() was called. 

> This is because there's also this snippet in writePump.pumpFinished which always calls tcpInputs.cancel() before calling tcpOutputs.complete:

Yes, but that is correct, if we are fullClosing, then if we finished with the write side we will cancel the read side, too, since we go full down. What the bug was here, is to send an Abort from the cancel which is completely unnecessary and wrong. 

But even fixing that, I have other issues, and when I saw "Monitored actor X terminated" in a DEBUG log, from IO actors, then I started to suspect a DeathPactException problem. And when I put a simple sleep in the tryShutdown() method, before the context.stop(), TcpSpec stopped to fail! It is still likely a bug to not unwatch the registered actor once we received a terminal close in the TCP actor.

So from my side the summary is:
 - There is a bug in tcpInputs.cancel() that it sends an Abort for no good reason, which will be triggered by halfClose=false (unconfirmed, will try out)
 - There is probably some other race, maybe the IO actor is DeathPacted after receiving the close (unconfifmed, trying to reproduce with raw IO and fix it)
 - There is my StressTest which only fails on my machine, never on Mathias's, but I suspect that the problem there is the enabled windows-connection-abort hack for that stresstest

So first I want to nail TcpSpec which has probably nothing with this ticket, and might be a lower layer IO bug. Once that is fixed, we can check if my removal of Abort in tcpInputs fixes the issue (it never fails on my machine unfortunately). Finally, I want the StressTest to not miserably fail on my machine

**Comment by jrudolph** [comment]
> Yes, but that is correct, if we are fullClosing, then if we finished with the write side we will cancel the read side, too, since we go full down. What the bug was here, is to send an Abort from the cancel which is completely unnecessary and wrong.

Yes, you are right. It must be the heat today or I'm still mentally on vacation...

**Comment by jrudolph** [comment]
> There is probably some other race, maybe the IO actor is DeathPacted after receiving the close (unconfifmed, trying to reproduce with raw IO and fix it)

I can confirm this related to the aborts for my reproduction. This is my explanation of what happens:

The problem always happens when the last write wasn't yet acked in the tcp stream actor, in this case:
   * it first sends a `ConfirmedClose` and waits for the ack
   * when the ack arrives it sends a `Close`
   * and then stops the stream actor

What the Akka.io actor sees in this case is this:
 * a `ConfirmedClose` either during the last write or after (it doesn't matter as it flushes ongoing writes correctly), after the last write has occurred it will call `shutdownOutput` and get into `closing` state
 * in this state it receives and ignores the `Close` command (it is unhandled, actually)
 * the handler goes away without the connection/channel being yet fullly closed: the connection is aborted in postStop

Several things are wrong in this protocol or need to be clarified:
 1. What should happen if the handler / close commander terminates in the `closing`, i.e. `ConfirmedClose` state?
 1. Should the sequence `ConfirmedClose` -> `Close` be valid?
 1. Is using `ConfirmedClose` the right thing to do while waiting for the last ack?

I'd say:

 1. In closing state (where we can still receive data), we should ignore if the close commander terminates and regularly close the connection if the handler terminates instead of aborting it. The reasoning is the subtle difference that the situation described in this ticket where a RST packet is sent in response to an expected ACK won't happen with a regular `channel.close` because of the `close` semantics described in the FAQ link above. In the case where the other side tries to send data it will still get a RST (can someone confirm this?) which is the same behavior as with the current `abort`.
 2. Yes, for simplicity reasons and to make the `Close` command idempotent, I'd say it makes sense to handle the `Close` command in the `closing` state and "upgrade" the half-close to a full-close, i.e. call `channel.close` and stop the `TcpConnection` actor.
 3. I don't think so (but correct me if I'm wrong). There's no reason to use `ConfirmedClose` just to wait for the last ack because `TcpConnection` correctly flushes the last write before executing any close command. Removing the "last write acking" logic from TCP streams should even simplify things.

Funny enough, I think, implementing any single one of the suggestions in the three answer would have prevented the situation described here.

**Comment by jrudolph** [comment]
I should maybe add that there's a notable absence of the solution of "just unwatching the handler / close commander", which would have the potential of leaving the connection in half-closed state if the peer silently goes away, right?

**Comment by drewhk** [comment]
Hmm. 

> it first sends a ConfirmedClose and waits for the ack
when the ack arrives it sends a Close

Are you sure that it first sends a ConfirmedClose? Because that is a bug, it should not send anything in fullClose mode, just the final Close after the final Ack. I think I missed there an if (halfClose)...

> I should maybe add that there's a notable absence of the solution of "just unwatching the handler / close commander", which would have the potential of leaving the connection in half-closed state if the peer silently goes away, right?

Well, I meant that unwatch is only needed on a "terminal" close operation, i.e. Close and Abort, but not ConfirmedClose (since the actor intends to still read and therefore it is an error to terminate it).

**Comment by jrudolph** [comment]
In the "terminal" close scenario it's only a problem if the closing is still blocked by a pending write, i.e. the actor is in the `closingWithPendingWrite` state. But you are right, that this needs to be considered as well.

> Are you sure that it first sends a ConfirmedClose? Because that is a bug, it should not send anything in fullClose mode, just the final Close after the final Ack. I think I missed there an if (halfClose)...

Here's the full code of `tcpOutputs.complete` I alluded to above:

```scala
    override def complete(): Unit = {
      if (!closed && initialized) {
        closed = true
        //println(s"tcpInputs.isClosed: ${tcpInputs.isClosed} halfClose: $halfClose lastWriteAcked: $lastWriteAcked")
        if (tcpInputs.isClosed && (halfClose || lastWriteAcked)) {
          // We can immediately close if
          // - half close mode, and read size already finished
          // - full close mode, and last write has been acked
          //
          // if in full close mode, and has a non-acknowledged write, we will do the closing in handleWrite
          // when the Ack arrives
          closeConnection(Close)
          tryShutdown()
        } else
          closeConnection(ConfirmedClose)
      }
    }
```

So, it seems it was a conscious decision which may need to be reconsidered now.

**Comment by drewhk** [comment]
No, it was just me being stupid. The idea was that on fullClose we should either:
 - immediately Close if there was no write for which we are waiting an Ack
 - Close after the final Ack arrived. (it's in handleWrite) 

I forgot that the else case can be triggered by (!halfClose && !lastWriteAcked). Only halfClose should send a ConfirmedClose there, so I need to add an extra if there.

**Comment by jrudolph** [comment]
Isn't the condition just wrong (it doesn't even implement the comment below)? What about just replacing it with

```scala
if (tcpInputs.isClosed || fullClose) {
```

**Comment by drewhk** [comment]
The idea was that if we are halfClose, but we already cancelled reading, or remote side closed write already (halfClose = true, tcpInputs.isClosed = true), then we just send a Close. Let's write a table to clear this up (and for both of us to review):

 - halfClose, tcpInputs.Closed - we were the only ones writing, or we were already uninterested in reading, so Close
 - halfClose, !tcpInputs.Closed - we stopped writing, but we are still interested in reading, so ConfirmedClose
 - !halfClose, !lastWriteAcked - we stopped writing, will stop after final Ack. send nothing. Please note that the writePump closes the tcpInputs at this point because of fullClose, so tcpInputs.isClosed is true
 - !halfClose, lastWriteAcked - we stopped writing, last write has been Acked, so Close. (again, tcpInputs.isClosed is already true, since writePump closed it, see above)

**Comment by jrudolph** [comment]
I agree with your table, I would have proposed in `!halfClose, !lastWriteAcked` to send the Close directly, and then wait or don't wait for the Ack to shutdown the stream actor. If you don't wait you could possibly lose a potential error that could still be dispatched on the input side and it would send the Ack to DeadLetters.

**Comment by jrudolph** [comment]
Would this work as a fix:

```scala
       if (tcpInputs.isClosed || fullClose) {
          closeConnection(Close)
          // in full close mode, we shut down in handleWrite only when the Ack arrives
          If (lastWriteAcked) tryShutdown()
        } else
          closeConnection(ConfirmedClose)
```

and 

```scala
    def handleWrite: Receive = {
      case WriteAck ⇒
        lastWriteAcked = true
        if (fullClose && closed) {
          // shudown after the last write has been flushed in full close mode.
          tryShutdown()
        }
        writePump.pump()
    }
```

In the `fullClose` case you could also delay sending the `Close` only in the last `handleWrite` (like before but without sending a `ConfirmedClose` before).

**Comment by jrudolph** [comment]
> In the fullClose case you could also delay sending the Close only in the last handleWrite (like before but without sending a ConfirmedClose before).

This alone suffices to fix the problem described in this ticket. Then, not even the change to Akka IO is needed:

```scala
else if (halfClose)
  closeConnection(ConfirmedClose)
```

i.e. only `if (halfClose)` needs to be added to prevent ever sending a `ConfirmedClose` in `fullClose` mode.

**Comment by drewhk** [comment]
I factored it a bit differently now, trying to avoid too much boolean logic and just write more explicit "if" branches.

**Comment by drewhk** [comment]
see: https://github.com/akka/akka/pull/17917

**Comment by ktoso** [comment]
After applying the patches: https://github.com/akka/akka/pull/17916 and https://github.com/akka/akka/pull/17917 we verified this on the same setup as in the original reproducer (versions, mac, java version) and we believe this to be fixed, output:

```
$ ab -c 1 -n 10000 http://127.0.0.1:8080/
This is ApacheBench, Version 2.3 <$Revision: 1604373 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 127.0.0.1 (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        akka-http/2.3-SNAPSHOT
Server Hostname:        127.0.0.1
Server Port:            8080

Document Path:          /
Document Length:        2 bytes

Concurrency Level:      1
Time taken for tests:   31.782 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      1690000 bytes
HTML transferred:       20000 bytes
Requests per second:    314.64 [#/sec] (mean)
Time per request:       3.178 [ms] (mean)
Time per request:       3.178 [ms] (mean, across all concurrent requests)
Transfer rate:          51.93 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0      10
Processing:     1    2   6.3      2     151
Waiting:        0    2   6.1      2     151
Total:          2    3   6.3      2     152

Percentage of the requests served within a certain time (ms)
  50%      2
  66%      2
  75%      3
  80%      3
  90%      3
  95%      3
  98%      4
  99%      9
 100%    152 (longest request)
```

**Comment by drewhk** [comment]
I also did a re-run of my previous stress-test: https://gist.github.com/drewhk/474651c62ea8ca003af6

It works fine now.

**Comment by viktorklang** [comment]
Excellent news!

On Tue, Jul 7, 2015 at 2:42 PM, drewhk <notifications@github.com> wrote:

> I also did a re-run of my previous stress-test:
> https://gist.github.com/drewhk/474651c62ea8ca003af6
>
> It works fine now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/issues/17854#issuecomment-119192362>.
>



-- 
Cheers,
√

**Comment by jrudolph** [comment]
Yay!

**Comment by drewhk** [comment]
@jrudolph Can you verify it just to be sure?

**Comment by jrudolph** [comment]
Yep, I tried it and found no problems. :+1:

**Comment by drewhk** [comment]
thanks!
